accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+In November of last year alone, we saw three supply chain attacks against popular npm packages ([UAParser.js](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack), [Command-Option-Argument](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack), and [rc](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack)), each with millions of downloads per month.
+
+The following chart illustrates several attack vectors facing the supply chain (src: [SLSA](https://slsa.dev/)):
+
+![SLSA: Supply-chain-threats](https://user-images.githubusercontent.com/20165/174750708-2be483ac-7e41-4bc3-8ee9-440ef33d9423.svg)
+
+Our focus is on mitigating package hijacking attacks where packages are built from a modified source (C) in the above diagram. This is where a malicious version of an existing open source package gets uploaded to the registry. These maliciously crafted versions are rarely reflected in the canonical source code repository and uploaded directly from an attacker's machine.
+
+The [above](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack) [attacks](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack) [are](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack) examples of this, and they frequently occur due to compromised npm credentials but can also happen due to compromised CI/CD or builds.
+
+Code on GitHub, GitLab etc can be browsed and audited, but packages in the registry are opaque and much harder to vet. If packages are built and published in the open these attacks become a whole lot harder. However, right now there’s no way of knowing where a package came from when you retrieve it from the registry.
+
+We want to solve the problem of npm packages being disconnected from their source by linking the published package to the source code repository and build that it originated from.
+
+Once a package includes provenance information (where and how it was built) we can start showing this when browsing a package on the npm registry. You would be able to click through to the particular commit and build that published a given version:
+
+<img width="376" alt="screenshot of npm package with provenance information" src="https://user-images.githubusercontent.com/20165/176637201-fdb02c11-c810-48e1-a203-ea4fb2008ad3.png">
+
+Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.
+
+Open source maintainers must be able to add build provenance information to their packages with near-zero initial and ongoing overhead.
```

Comment:
Given that local builds and publishes are not in-scope, this language should be toned down unless there's an "easy button" somewhere. 

If I'm understanding correctly, maintainers will have to change their publishing workflows to run in a "trusted" cloud environment. Maybe I'm out of the loop, but with TOTP enabled this has not historically been trivial to  set up. 

IMO this seems like a considerable barrier to adoption, at least today.

Reply:
@boneskull yes you're right, the main barrier will be changing publish workflows to run on supported CI/CD system. But beyond running the publish in a supported env, nothing else should need to be maintained or done, e.g. managing keys etc. I'll see how I can rephrase this. 

We're also planning on supporting publishing via OIDC in future which should make using CI/CD a lot more appealing. The setup for linking your source and build and publishing via OIDC would be identical.

So instead of using long-lived access tokens, or MFA/TOTP you could instead configure which repo/workflow is allowed to publish what package. The CLI would then exchange a OIDC id token from the CI/CD system for a short-lived access token which is used to publish the package.

Reply:
It's also worth mentioning that automation tokens work today for publishing via CI when you have 2FA enabled on your accounts.

Reply:
Yes as part of this work we'll want to make sure it's easy to move publishing to CI/CD by providing a reusable workflow/trusted builder that supports common publish workflows. Hopefully good documentation and reusable publish workflows will alleviate some of the initial lift in moving publishing to CI/CD.

Reply:
Automation tokens are a single factor, though.

Reply:
Often when discussing multiple factors it is in the context of Authentication (the A in MFA). It is definitely important to have multiple factors used when creating a new session, and nothing about that is changed by utilizing an automation token. While it is a “single factor” credential, we have not historically seen ATO or project compromise due to access token theft, it has traditionally been from leaked passwords which were in turn then used to mint access tokens. The introduction of login verification and enforcement of 2FA (for authentication) of high-impact projects has all been focused on securing authentication primarily.

We do plan to continue to invest in making more secure practices for package publishing. Next quarter we will be kicking off a private beta for granular access tokens which will allow you to limit tokens to publish specific packages or  scopes, set expiration dates, and limit to specific CIDR ranges (only can publish from certain IPs). This will significantly improve the baseline security of using tokens in CI.

Longer term, we are exploring introducing OIDC as a publish mechanism which would allow publishing between a CI environment like actions and npm without the use of any long lived tokens. This would come coupled with experience improvements to make setting this process up on GitHub seamless, while laying an open foundation for other providers to implement similar integrations.

We’ve extensively reviewed these solutions with our security experts at GitHub and believe that they offer adequate security to projects, if they didn’t we would not make these features available. One can always be more secure, and I don’t say that to discourage folks from having excellent security hygiene (I have a yubikey on my keychain), but there does become a balance between usability and security and often a compromise is necessary. It is my personal opinion that the trade of of introducing a token and using CI to publish it worth the security benefits that we get knowing each publish is coming from a fresh machine.